src/main.rs

```diff
@@ -1,54 +1,121 @@
-#![allow(dead_code)]
-extern crate rowan;
-use std::{error::Error, fs};
+mod parser;
 mod user;
+
+use std::{fs, str::FromStr};
+
+use parser::kill_node;
 use rnix::types::*;
 use user::*;
-mod parser;
-use parser::kill_node;
 
-fn main() -> Result<(), Box<dyn Error>> {
+struct ActionStack {
+    inner: Vec<UserAction>,
+}
+
+impl ActionStack {
+    fn new() -> Self {
+        Self {
+            inner: vec![UserAction::Intro],
+        }
+    }
+
+    fn push(&mut self, action: UserAction) {
+        self.inner.push(action)
+    }
+
+    fn pop(&mut self) -> UserAction {
+        if self.inner.len() > 1 {
+            self.inner.pop().unwrap()
+        } else {
+            *self.inner.last().unwrap()
+        }
```

Comment:
It irritates me that the behavoir of `push` vs `pop` is blatantly unsymetrical without any comment explaining why.
why not just `self.inner.pop().unwrap()` without the if? or:
```suggestion
        let ret = self.inner.pop().unwrap();
        if self.inner.is_empty() {
            self.clear();
        }
```

Reply:
The suggestion will panic if the inner `Vec` is empty; the if is there to ensure the `Vec` is never empty. The purpose is to always keep the first item in the stack (which will be `UserAction::Intro`). This way we can ensure that the stack can't be used wrongly, and all `pop`s must eventually come to the first `Intro` menu.